example/particles/particles.cpp

```diff
@@ -316,39 +318,176 @@ TaskStatus Defrag(MeshBlock *pmb) {
   return TaskStatus::complete;
 }
 
-// See the advection_driver.hpp declaration for a description of how this function gets
-// called.
-TaskCollection ParticleDriver::MakeTaskCollection(BlockList_t &blocks, int stage) {
+// Custom step function to allow for looping over MPI-related tasks until complete
+TaskListStatus ParticleDriver::Step() {
+  TaskListStatus status;
+  integrator->dt = tm.dt;
+
+  BlockList_t &blocks = pmesh->block_list;
+  auto num_task_lists_executed_independently = blocks.size();
+
+  // Create all the particles that will be created during the step
+  TaskCollection ptc = MakeParticlesCreationTaskCollection();
+  ptc.Execute();
```

Comment:
How much of this machinery we can provide internal to the infrastructure is something I've wondered about... At the moment it seems like we're asking the app dev to do way more work than necessary. This is nastiest in `MakeParticlesUpdateTaskCollection`, because while `Send` and `Receive` are infrastructure methods, one needs to write out a `Transport` function explicitly and included it with those other tasks. I'm not sure how to get around this... One possibility would be letting Parthenon craft its own `Transport` function by instead having the app dev enroll transport "events", like `crossed_a_cell`, `moved_a_finite_time`, `absorbed_by_a_piece_of_paper`, etc., like `dt_cell`, `dt_end` in the particles example. Then, depending on which event occurs each push, call the appropriate `do_some_physics_or_mark_the_particle_complete` or whatever. This is more complicated and potentially less flexible, but it might also be really nice.

Reply:
Also yes `MakeParticlesCreationTaskCollection().Execute();` works and is nicer, thanks

Reply:
> One possibility would be letting Parthenon craft its own Transport function by instead having the app dev enroll transport "events"

Interesting. That might be a nice idea in general for iterative methods. It might also make designing a flow control really difficult though. I'm not sure.

Reply:
I think the flow would be unaffected since the extra functionality you're adding is inside each particle's transport cycle, e.g.
```c++
pmb->par_for(KOKKOS_LAMBDA(const int n) {
  if (mask(n)) {
   while (t(n) < t0 + dt) {
      // Calculate delta time to each custom transport "event" enrolled by the app

      // Do the event that occurs at the shortest delta time, then update t(n) by that delta time
    }
  }
});
```
so it doesn't touch the tasking at all

Reply:
This is flow control though to some degree. Still I think it makes sense.